src/features/device-state/routes/state-patch-v3.ts

```diff
@@ -0,0 +1,384 @@
+							releases,
+						)) {
+							appReleaseUuids[appUuid].add(releaseUuid);
+							if (services != null) {
+								for (const [, service] of Object.entries(services)) {
+									imageLocations.push(service.image);
+								}
+							}
+						}
+					}
+				}
+			}
+		}
+
+		let devices: Array<
+			Pick<Device, 'id' | 'uuid'> & {
+				belongs_to__application: Array<Pick<Application, 'uuid'>>;
+			}
+		>;
+		let images: Array<Pick<Image, 'id' | 'is_stored_at__image_location'>>;
+		const appReleases: {
+			[appUuid: string]: Array<Pick<Release, 'id' | 'commit'>>;
+		} = {};
+		await sbvrUtils.db.readTransaction(async (tx) => {
+			const resinApiTx = api.resin.clone({
+				passthrough: { req, custom, tx },
+			});
+			devices = (await resinApiTx.get({
+				resource: 'device',
+				options: {
+					$select: ['id', 'uuid', 'belongs_to__application'],
+					$filter: {
+						uuid: { $in: uuids },
+					},
+					$expand: {
+						belongs_to__application: {
+							$select: 'uuid',
+						},
+					},
+				},
+			})) as Array<
+				Pick<Device, 'id' | 'uuid'> & {
+					belongs_to__application: Array<Pick<Application, 'uuid'>>;
+				}
+			>;
+			if (devices.length !== uuids.length) {
+				throw new UnauthorizedError();
+			}
+
+			if (imageLocations.length > 0) {
+				const imgLocationFilter = imageLocations.map((imgLocation) => {
+					const [location, contentHash] = imgLocation.split('@');
+					const filter: Filter = { is_stored_at__image_location: location };
+					if (contentHash) {
+						filter.content_hash = contentHash;
+					}
+					return filter;
+				});
+				images = (await resinApiTx.get({
+					resource: 'image',
+					options: {
+						$select: ['id', 'is_stored_at__image_location'],
+						$filter:
+							imgLocationFilter.length === 1
+								? imgLocationFilter[0]
+								: imgLocationFilter,
+					},
+				})) as Array<Pick<Image, 'id' | 'is_stored_at__image_location'>>;
+				if (imageLocations.length !== images.length) {
+					throw new UnauthorizedError();
+				}
+			}
+
+			for (const [appUuid, releaseUuids] of Object.entries(appReleaseUuids)) {
+				appReleases[appUuid] =
+					releaseUuids.size === 0
+						? []
+						: ((await resinApiTx.get({
+								resource: 'release',
+								options: {
+									$select: ['id', 'commit'],
+									$filter: {
+										commit: { $in: Array.from(releaseUuids) },
+										belongs_to__application: {
+											$any: {
+												$alias: 'a',
+												$expr: {
+													a: { uuid: appUuid },
+												},
+											},
+										},
+									},
+								},
+						  })) as Array<Pick<Release, 'id' | 'commit'>>);
+				if (appReleases[appUuid].length !== releaseUuids.size) {
+					throw new UnauthorizedError();
+				}
+			}
+		});
+
+		await sbvrUtils.db.transaction(async (tx) => {
+			const resinApiTx = api.resin.clone({
+				passthrough: { req, custom, tx },
+			});
+
+			const waitPromises: Array<PromiseLike<any>> = [];
+
+			for (const uuid of uuids) {
```

Comment:
Do we need to respect the order in which the device UUIDs were send?
If not, how about iterating over the retrieved devices directly (since we already checked that the lengths do match), so that we don't have to use `const device = devices.find((d) => d.uuid === uuid);` later ?
```suggestion
			for (const device of devices) {
```
Even if the order does matter, could a `$orderby: { depends_on__device: 'asc' }` work around it?

Reply:
I don't think the order matters, and regardless I'd only actually expect a single one to be sent right now anyway, so I'll change this